DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/optimize_inner_loop/src/optimize_inner_loop.cpp

```diff
@@ -0,0 +1,280 @@
+  static_assert(Version >= 0, "Invalid kernel version");
+  static_assert(SpecIters >= 0, "SpecIters must be positive");
+
+  // the pipe
+  using Pipe = pipe<PipeClass<Version>, bool>;
+
+  kernel_time_ms = 0.0;
+  int size = in.size();
+
+  try {
+    // create the device queue with profiling enabled
+    auto prop_list = property_list{ property::queue::enable_profiling() };
+    queue q(selector, dpc_common::exception_handler, prop_list);
+
+    // The input data buffer
+    buffer in_buf(in);
+
+    // The output data buffer
+    // Scalar inputs are passed to the kernel using the lambda capture,
+    // but a SYCL buffer must be used to return a scalar from the kernel.
+    buffer<int, 1> res_buf(&res, 1);
+
+    // submit the Producer kernel
+    event p_e = q.submit([&](handler &h) {
+      // the input buffer accessor
+      accessor in_accessor{ in_buf, h, read_only };
+
+      h.single_task<Producer<Version>>([=]() [[intel::kernel_args_restrict]] {
+        for (int i = 0; i < size; i++) {
+          // read the input value, which is in the range [0,InnerLoopBound]
+          int val = in_accessor[i];
+
+          // 'InElementUpperBound' is a constant (a template variable).
+          // Therefore, the condition 'InElementUpperBound < 0', and therefore
+          // the taken branch of this if-else statement, can be determined at
+          // compile time. This results in the branch that is NOT taken being
+          // optimized away. Both versions of the inner loop apply the
+          // speculated_iterations attribute, where the number of speculated
+          // iterations is determined by the template variable 'SpecIters'.
+          if (InElementUpperBound < 0) {
+            // In this version of the inner loop, we do NOT provide an
+            // upperbound on the loop index variable 'j'. While it may be easy
+            // for you to read the code and reason that 'j<InElementUpperBound'
+            // is always true by looking at the rest of the program, it is much
+            // more difficult for the compiler. As a result, the compiler will
+            // be conservative and assume this inner loop may have a large trip
+            // count and decide to make (or not make) optimizations accordingly.
+            [[intelfpga::speculated_iterations(SpecIters)]]
+            for (int j = 0; j < val; j++) {
+              Pipe::write(true);
+            }
+          } else {
+            // In this version of the inner loop, we provide an upper bound
+            // on the loop index variable 'j' by adding the
+            // 'j<InElementUpperBound' loop exit condition. This provides the
+            // compiler with a constant upperbound on the trip count and allows
+            // it to make optimizations accordingly.
+            [[intelfpga::speculated_iterations(SpecIters)]]
+            for (int j = 0; j < val && j <= InElementUpperBound; j++) {
```

Comment:
Interesting...
Could your compiler also infer the the upper bound with:
```suggestion
            if (val <= InElementUpperBound + 1)
              for (int j = 0; j < val; j++) {
```
if so I would be interested by knowing whether there is a performance improvement, specially on host device or emulation mode.

Reply:
Hmmm not sure what you are suggesting here based on the 'suggested change'.

> if so I would be interested by knowing whether there is a performance improvement, specially on host device or emulation mode.
Why does host device/emulation matter? This is a FPGA specific optimization. Saying that, maybe there are instances where CPUs do better when the trip counts are low and bound too.

Reply:
I suggest hoisting one of the test out of the loop https://en.wikipedia.org/wiki/Loop-invariant_code_motion and see if there is an impact on the performance and the QoR.
If it works, the side effect is that this FPGA-friendly code can work better on CPU too.
Even if you do not care, the SYCL philosophy is that just running this FPGA-targeted code on plain CPU should run faster than using FPGA emulation, as a friendly way to debug SYCL application...

Reply:
@aisoard can we extract this precondition in both cases?

Reply:
Ohhhhh sorry, I thought you were suggesting to change the design. 
You are asking if the compiler can infer the maximum loop bounds on your new piece of code. And if so, what type of performance changes we see for the CPU as well. Is this correct?

Reply:
To answer those in the short term:

> Can the compiler infer the maximum loop bounds on your new piece of code?

Yes

> Performance improvements for the CPU

For another day...

Reply:
No problem. Just curious whether we can find a coding style which is good for FPGA but still OK on CPU. :-)

Reply:
What is a CPU :S ???